File: src/main/java/com/github/theonlytails/rubymod/registries/TileEntityTypes.java
package com.github.theonlytails.rubymod.registries;

import com.github.theonlytails.rubymod.RubyMod;
import com.github.theonlytails.rubymod.tileentities.RubyBarrelTileEntity;
import net.minecraft.tileentity.TileEntityType;
import net.minecraftforge.fml.RegistryObject;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.ForgeRegistries;

public class TileEntityTypes {
    public static final DeferredRegister<TileEntityType<?>> TILE_ENTITIES = DeferredRegister.create(ForgeRegistries.TILE_ENTITIES, RubyMod.MOD_ID);

    @SuppressWarnings("ConstantConditions")
    public static final RegistryObject<TileEntityType<RubyBarrelTileEntity>> RUBY_BARREL = TILE_ENTITIES.register(
            "ruby_barrel",
            () -> TileEntityType.Builder
                    .create(RubyBarrelTileEntity::new,
                            BlockRegistry.RUBY_BARREL.get())
                    .build(null)
    );
}

File: src/main/kotlin/com/github/theonlytails/rubymod/registries/TileEntityTypes.kt
package com.github.theonlytails.rubymod.registries

import com.github.theonlytails.rubymod.RubyMod
import com.github.theonlytails.rubymod.tileentities.RubyBarrelTileEntity
import net.minecraft.tileentity.TileEntityType
import net.minecraftforge.registries.ForgeRegistries
import thedarkcolour.kotlinforforge.forge.KDeferredRegister

/**
 * Registers custom tile entities.
 *
 * @author TheOnlyTails
 */
object TileEntityTypes {
	val TILE_ENTITIES = KDeferredRegister(ForgeRegistries.TILE_ENTITIES, RubyMod.MOD_ID)

	val RUBY_BARREL: TileEntityType<RubyBarrelTileEntity> by TILE_ENTITIES.register("ruby_barrel") {
		@Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
		TileEntityType.Builder
			.create(::RubyBarrelTileEntity, BlockRegistry.RUBY_BARREL)
			.build(null)
	}
}
